### PR TITLE
Support Map Type

### LINKED ISSUE #56: Support Map type
The AssemblyScript [`Map`](https://www.assemblyscript.org/stdlib/map.html) type appears to not be supported.

For example:

```json
{
  "data": {
    "a": "x",
    "b": "y",
    "c": "z"
  }
}
```

should be deserializable to:

```ts
@json
class Container {
    data!: Map<string, string>;
}
```

But I get an error such as:

> Could not deserialize data {"a":"x","b":"y","c":"z"} to type Map<~lib/string/String,~lib/string/String>. Make sure to add the correct decorators to classes.

### PR DESCRIPTION
**Pull request by mattjohnsonpint**
Resolves #56

### CODE REVIEW

File: assembly/src/json.ts [hunk 1]
```diff
@@ -725,3 +905,8 @@ function parseDate(dateTimeString: string): Date {
   // is globally aliased to wasi_Date (or some other superclass).
   return new Date(d.getTime());
 }
+
+// @ts-ignore: Decorator
+@inline function isMap<T>(): bool {
+  return nameof<T>().startsWith("Map<");
```
Comment (mattjohnsonpint) [inline assembly/src/json.ts:911]: @JairusSW - Happy to change this if you can think of a better way.  Thanks.

File: assembly/src/json.ts [hunk 2]
```diff
@@ -559,6 +592,153 @@ export namespace JSON {
...
+      }
+    }
+  }
+
+  return map;
+}
+
+//@ts-ignore: Decorator
+@inline function parseMapKey<T>(key: Virtual<string>): T {
+  const k = key.copyOut();
```
Comment (mattjohnsonpint) [inline assembly/src/json.ts:729]: @JairusSW - is `copyOut` the best way to handle this?  And does that defeat the purpose of it being virtual to begin with?  If there's a better way, please let me know.  Thanks.

**Comment by mattjohnsonpint** [comment]
will fix CI failure shortly...

File: assembly/src/json.ts [hunk 3]
```diff
@@ -559,6 +592,156 @@ export namespace JSON {
   return schema;
 }
 
+// @ts-ignore: Decorator
+@inline function parseMap<T extends Map>(data: string, initializeDefaultValues: boolean): T {
+
+  const map: nonnull<T> = changetype<nonnull<T>>(
+    __new(offsetof<nonnull<T>>(), idof<nonnull<T>>())
+  );
```
Comment (mattjohnsonpint) [inline assembly/src/json.ts:600]: Interestingly, `const map = instantiate<T>()` worked just fine for `yarn run test`, but failed to compile when calling `yarn run build:test`.  Not sure why, but this alternate approach works for both.

Also, FYI - there remiain three warnings when calling `yarn run build:test` that are unrelated to this PR.

File: assembly/src/json.ts (hunk 1 above)
Comment (JairusSW) [reply assembly/src/json.ts:914]: I'd use some functionality like this to do that at compile time
```js
// @ts-ignore: Decorator
@inline function isMap<T>(): boolean {
    let type = changetype<T>(0);
    return type instanceof Map;
}
```
Its pretty wack, but it works well

File: assembly/src/json.ts (hunk 2 above)
Comment (JairusSW) [reply assembly/src/json.ts:732]: Using `copyOut()` here is going to be your best option. That is, unless I overhaul each distinct parser to accept `Virtual<T>` as input, which is not the case at the moment. `copyOut()` is appropriate here.

File: assembly/__tests__/as-json.spec.ts [hunk 4]
```diff
@@ -202,4 +202,148 @@ describe("Deser externals", () => {
         body: '{\n  "args": {},\n  "headers": {\n    "content-length": "0",\n    "accept": "*/*" \n}}'
       })
   })
-});
\ No newline at end of file
+});
```
Comment (JairusSW) [inline assembly/__tests__/as-json.spec.ts:205]: Tests look good here

File: assembly/src/json.ts (hunk 3 above)
Comment (JairusSW) [reply assembly/src/json.ts:600]: I checked on my end and it seems that those warnings are a result of using a newer AssemblyScript version. I'll update accordingly.
That's quite interesting. Not sure why thats failing to build

**Comment by JairusSW** [comment]
Hey, @mattjohnsonpint!
Thanks for your contribution, I appreciate it.
I left a few comments, so let me know when you feel good about this PR and I'm willing to merge it into main.

Please add yourself to the contributors list in `package.json` if not already done.

File: assembly/src/json.ts (hunk 1 above)
Comment (mattjohnsonpint) [reply assembly/src/json.ts]: Perfect.  Thanks!

File: assembly/src/json.ts (hunk 2 above)
Comment (mattjohnsonpint) [reply assembly/src/json.ts:732]: That's fine. I can just use `copyOut` for now.  I think the perf is still reasonable.

**Comment by mattjohnsonpint** [comment]
Suggestions applied.  Please merge and release at your convenience.  Thanks!

**Comment by mattjohnsonpint** [comment]
Moved repo maintenance bits to #60.  Thanks.
